designs/2019-esm-compatibilty/README.md

```diff
@@ -0,0 +1,222 @@
+
+In excepitonal edge cases it may require extra work to establish interoperability across package boundaries. The way ESLint uses `.js` config files is one of those edge cases.
+
+The goal of this RFC is to explore the different options for addressing this issue.
+
+## Detailed Design
+
+To understand the design outlined here requires some background info into the mechanics by which Node modules are loaded.
+
+### Jargon
+
+- Package - A NPM package (ie contains `package.json`)
+- Module - A JS file
+- CJS - A CommonJS module
+- ESM - A Standard ECMAScript module
+- Consumer - A package that uses/depends on this package
+- Dependent - Package(s) that this package needs to work
+- Boundary - The demarcation between package, consumer, dependent
+
+### A Crash Course on Package Boundaries
+
+Historically, NPM packages have come in a variety of different package formats (ie IIFE, AMD, UMD, CJS). Prior to an actual spec, NPM settled on CJS as it's de-facto module format. CJS isn't going anywhere, the addition of ESM is additive.
+
+By default, all NPM packages are CJS-based. That means all `.js` files will be read as CJS modules. To include an ES module in a CJS package, it must have the `.mjs` extension.
+
+In addition, if `package.json` contains `"type": "module"` then the package is ESM-based. Meaning, all `.js` files contained within will be treated as ESM. To include a CJS module in a ESM-based package, it must have the `.cjs` extension.
+
+This configuration option does *not* affect consumers or dependents. Their functionality depends entirely on their own configuration. Assuming all packages follow this rule, there should be no issues with interoperability between CJS/ESM.
+
+### The Scenario
+
+A user is creating a new package. They want it to be ESM for Browser <-> Node compatibility so they configure their package to be ESM-based.
+
+The user adds `eslint` as a devDependency, w/ a typical NPM script to lint the package contents.
+
+The user defines `eslint.config.js` outlining the ruleset they prefer to use within their package.
+
+### The Issue
+
+The user pacakge is ESM-based, all `.js` files within are read as ESM.
+
+ESLint is CJS-based, it loads all files within it's package boundary as CJS.
+
+The configuration file is defined as a CJS module (ie `module.exports`), but has the `.js` syntax so requiring it throws an error. ESLint, by design reaches across the package boundary to load the configuration but the user has inadvertently signaled to Node that it uses the wrong module type.
+
+### The Fix
+
+Rename `eslint.config.js -> eslint.config.cjs`
+
+There is no documentation outlining how to define a ESM-based configuration. As such, it can be assumed that any JS-based ESLint config will always be a CJS module.
+
+Therefore, the `.cjs` extension needs to be assiciated w/ JS-based configrations.
+
+*/lib/cli-engine/config-array-factory.js*
+
+```diff
+function loadConfigFile(filePath) {
```

Comment:
This function doesn't affect how ESLint finds config files. We should update the finding logic.

Reply:
I could use some help w/ this, I'm not very familiar w/ the core internals other than what I've been able to find so far.

Reply:
Actually, RFCs don't need to include implementation :)

For a reference, [this array](https://github.com/eslint/eslint/blob/master/lib/cli-engine/config-array-factory.js#L52-L59) is the order to find config files. ESLint checks files in the array one by one then adopts the first found one.

Therefore the current priority is:

1. `.eslintrc.js`
1. `.eslintrc.yaml`
1. `.eslintrc.yml`
1. `.eslintrc.json`
1. `.eslintrc`
1. `package.json`

This RFC should include how it changes this priority because it's a user-facing change.

Reply:
And the priority is documented in the "[Configuration File Formats](https://eslint.org/docs/user-guide/configuring#configuration-file-formats)" section. The "Documentation" section in this RFC should include it.

Reply:
The required change would put `.eslintrc.cjs` at priority 1. In short, it's an 'escape hatch' that will primarily impact users who use `"type": "module"` **and** `.eslintrc` style configs.

I'll see if I can clarify the descriptions and reserve implementation specifics for the related PR.

Reply:
In this case it'll become:

1. .eslintrc.cjs
2. .eslintrc.js
3. .eslintrc.yaml
4. .eslintrc.yml
5. .eslintrc.json
6. .eslintrc
7. package.json

The implementation specifics have been removed in the RFC, this change will be added to the PR instead.

Reply:
Would you include the priority change in this RFC? And would you add about we need to update the "[Configuration File Formats](https://eslint.org/docs/user-guide/configuring#configuration-file-formats)" section to the Documentation section of this RFC?

Reply:
Done. See Revision 5.